File: 0143-reorder-list/0143-reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    // Approach: store the node references in the array, and then link the nodes by their array index
    // Time: O(n)
    // Space: O(n)

    void reorderList(ListNode* head) {
        vector<ListNode*> nodes;
        ListNode* curr = head;
        while (curr) {
            nodes.push_back(curr);
            curr = curr->next;
        }

        int start = 0, end = nodes.size() - 1;
        while (start <= end) {
            // handle cases of odd number of nodes
            if (start == end) {
                nodes[start]->next = nullptr;
                break;
            }
            // handle cases of even number of nodes
            if (start + 1 == end) {
                nodes[start]->next = nodes[end];
                nodes[end]->next = nullptr;
                break;
            }
            
            nodes[start]->next = nodes[end];
            nodes[end]->next = nodes[start + 1];

            start++;
            end--;
        }
    }
};

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head or not head.next:
            return

        # separate the left and right sides
        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next

        right_head = slow.next
        slow.next = None

        # reverse the right linked list
        prev = None
        curr = right_head
        while curr:
            tmp = curr.next
            curr.next = prev
            prev = curr
            curr = tmp
        right_head = prev

        # merge
        left_curr = head
        right_curr = right_head
        while right_curr:
            left_next = left_curr.next
            right_next = right_curr.next
            left_curr.next = right_curr
            right_curr.next = left_next

            left_curr = left_next
            right_curr = right_next


# separate
# 1 2 3
# 4 5

# inverse
# 5 4

# merge
# 1 2 3
# 5 4

# 1 5 2 4 3